lib/config/applyrule-targeted.cpp

```diff
@@ -0,0 +1,280 @@
+
+		if (GetTargetServices(rule.m_Filter.get(), services)) {
+			ApplyRule::Ptr sharedRule = new ApplyRule(std::move(rule));
+			sharedRule->m_IsTargeted = true;
+
+			for (auto service : services) {
+				rules.Targeted[*service.first].ForServices[*service.second].emplace_back(sharedRule);
+			}
+
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/**
+ * If the given assign filter is like the following, extract the host names ("H", "h", ...) into the vector:
+ *
+ * host.name == "H" [ || host.name == "h" ... ]
+ *
+ * The order of operands of || == doesn't matter.
+ *
+ * @returns Whether the given assign filter is like above.
+ */
+bool ApplyRule::GetTargetHosts(Expression* assignFilter, std::vector<const String *>& hosts)
+{
+	auto lor (dynamic_cast<LogicalOrExpression*>(assignFilter));
+
+	if (lor) {
+		return GetTargetHosts(lor->GetOperand1().get(), hosts)
+			&& GetTargetHosts(lor->GetOperand2().get(), hosts);
+	}
+
+	auto name (GetComparedName(assignFilter, "host"));
+
+	if (name) {
+		hosts.emplace_back(name);
+		return true;
+	}
+
+	return false;
+}
+
+/**
+ * If the given assign filter is like the following, extract the host+service names ("H"+"S", "h"+"s", ...) into the vector:
+ *
+ * host.name == "H" && service.name == "S" [ || host.name == "h" && service.name == "s" ... ]
+ *
+ * The order of operands of || && == doesn't matter.
+ *
+ * @returns Whether the given assign filter is like above.
+ */
+bool ApplyRule::GetTargetServices(Expression* assignFilter, std::vector<std::pair<const String *, const String *>>& services)
+{
+	auto lor (dynamic_cast<LogicalOrExpression*>(assignFilter));
+
+	if (lor) {
+		return GetTargetServices(lor->GetOperand1().get(), services)
+			&& GetTargetServices(lor->GetOperand2().get(), services);
+	}
+
+	auto service (GetTargetService(assignFilter));
+
+	if (service.first) {
+		services.emplace_back(service);
+		return true;
+	}
+
+	return false;
+}
+
+/**
+ * If the given filter is like the following, extract the host+service names ("H"+"S"):
+ *
+ * host.name == "H" && service.name == "S"
+ *
+ * The order of operands of && == doesn't matter.
+ *
+ * @returns {host, service} on success and {nullptr, nullptr} on failure.
+ */
+std::pair<const String *, const String *> ApplyRule::GetTargetService(Expression* assignFilter)
+{
+	auto land (dynamic_cast<LogicalAndExpression*>(assignFilter));
+
+	if (!land) {
+		return {nullptr, nullptr};
+	}
+
+	auto op1 (land->GetOperand1().get());
+	auto op2 (land->GetOperand2().get());
+
+	{
+		auto host (GetComparedName(op1, "host"));
+
+		if (host) {
+			auto service (GetComparedName(op2, "service"));
+
+			if (!service) {
+				return {nullptr, nullptr};
+			}
+
+			return {host, service};
+		}
+	}
+
+	auto service (GetComparedName(op1, "service"));
+
+	if (!service) {
+		return {nullptr, nullptr};
+	}
+
+	auto host (GetComparedName(op2, "host"));
+
+	if (!host) {
+		return {nullptr, nullptr};
+	}
+
+	return {host, service};
+}
```

Comment:
This could employ a small trick to save some code duplication: Just do this:

```
auto host (GetComparedName(op1, "host"));
if (!host) {
    std::swap(op1, op2);
    host = GetComparedName(op1, "host");
}
```

And then you can handle both cases with the op1 is host and op2 is service case.